### PR TITLE
Performance investigation and improvements for `random_agent`.

### PR DESCRIPTION
**Pull request by jacobusmmsmit**
This is a draft PR intended for discussion and changes before pulling.

This PR addresses #742 by implementing two new functions for getting a random agent from an ABM as well as benchmarking them to find out which is most suitable. Additionally, the `random_agent` function is given a new kwarg `optimistic=false` controlling whether it should do a (normally) non-allocating "optimistic" search using `optimistic_random_agent` or the traditional method, renamed to `allocating_random_agent`.

`optimistic_random_agent` is implemented as follows: pick a random agent, does it satisfy the condition?. It does this until `n_attempts` is reached, at which point one of two things could potentially happen:
1. Fallback to (an implementation of) `allocating_random_agent` (as it currently does) and return a random agent.
2. Return nothing.

The former offers stronger correctness guarantees, the latter stronger performance guarantees (the function will never allocate, compilers like this, but it still is impure due to randomness so probably not such a big deal) while sacrificing correctness.

Additionally, this PR also implements what I claim is a more efficient `allocating_random_agent` (currently called `allocating_random_agent2`) which first applies an inplace filter to the collected agent IDs, then just picks one at random. Benchmarks can be found in `test/performance/get_random_agent.jl` and summarised below as:

```julia
# All times are median
# FixedMassABM
@benchmark optimistic_random_agent(fixed_model, cond) # 336 ns
@benchmark allocating_random_agent(fixed_model, cond) # 1450 ns
@benchmark allocating_random_agent2(fixed_model, cond) # 6100 ns

# UnkillableABM
@benchmark optimistic_random_agent(noremove_model, cond) # 158 ns 
@benchmark allocating_random_agent(noremove_model, cond) # 1283 ns
@benchmark allocating_random_agent2(noremove_model, cond) # 276 ns

# DictionaryABM
@benchmark optimistic_random_agent(dict_model, cond) # 454 ns
@benchmark allocating_random_agent(dict_model, cond) # 1592 ns
@benchmark allocating_random_agent2(dict_model, cond) # 882 ns
```

The results show that apart from for fixed-mass ABMs, the filter-first technique is an improvement. It allocates the same (number and size).

### CODE REVIEW

**Comment by jacobusmmsmit** [comment]
A key thing that I haven't mentioned is that it is very lenient condition that I am checking here. 50% of the agents have this condition satisfied. Speedup would be less for fewer agents.

**Comment by Tortar** [comment]
Cool idea to have a probabilistic search! I modified the old `allocating_random_agent`, now the ways are more or less of the same performance

**Comment by jacobusmmsmit** [comment]
If the performance difference between the two methods is consistently different between container types it might be worth dispatching on it.

**Comment by Tortar** [comment]
need to try one last optimization of `allocating_random_agent`, but I like the idea!

**Comment by Tortar** [comment]
On the other hand, with the optimistic search I don't know if the dispatching is worthy enough

**Comment by Tortar** [comment]
Also I'm finding a bit strange that the new way is so much better in one case, but worse in the others since the two functions both collects ids and so they then operate on the same data structures from that point on, while the `allocating_random_agent` is almost constant, so how can it be?

**Comment by Tortar** [comment]
wait a second, in the benchmark it shouldn't be

```
noremove_model = UnkillableABM(agents)
dict_model = ABM(agents)
fixed_model = FixedMassABM(agents)
```
?

**Comment by Tortar** [comment]
the new function should be much better than both! with the 10000 agents as above with 50% density I calculated

```julia
# FixedMassABM
6.265 μs
# UnkillableABM
6.066 μs 
# DictionaryABM
17.282 μs
```

Hope it's correct

**Comment by jacobusmmsmit** [comment]
That looks great, I had a go trying out `popat!` instead of swapping elements and using a decrementing `N` each time instead of calling `eachindex` each loop, but your method seems to be a few `ns` faster in each case.

**Comment by jacobusmmsmit** [comment]
Ok... I don't feel like I have the context to know what this broken test is trying to do. @Datseris ?

**Comment by Datseris** [comment]
Failure is related with a badly written test.

**Comment by Datseris** [comment]
Yoooooooooooooo what happened here, I merged main and your code is gonne :D haha sorry about that, please retrieve your code and put it at the appropriate new file!

**Comment by Tortar** [comment]
@jacobusmmsmit just to mention: `popat!` should become slower and slower than swapping and removing the last more and more the agent number gets greater

**Comment by Datseris** [comment]
Okay so what is the status quo here? Two new versions are offered, the old version was removed. Both versions are faster than the old?

**Comment by jacobusmmsmit** [comment]
Ok, I've reinstated the changes. I think the problem was that the merge from main pulled changes that deleted the file, as there are now 2 new files in 5.7: `model_abstract` and `model_concrete`.

**Comment by Tortar** [comment]
the last version has proved the fastest, that's why @jacobusmmsmit removed the others. The first version proposed by him was faster but not always, while this last one is (up to 20x) than the old one with all models.

**Comment by Datseris** [comment]
but the optimistic version is not the default, right?

**Comment by Datseris** [comment]
I admit I find it difficult to believe that a single `collect` and `shuffle!` is slower than `collect` and arbitrarily many calls to `rand` and `pop!`, but then again you tested it and the numnbers don't lie.

Please replace all calls to `model.rng` with `abmrng(mode)`, add a changelog entry and bump minor release up so that I can merge and tag this improvement !

**Comment by jacobusmmsmit** [comment]
Same lol

**Comment by jacobusmmsmit** [comment]
I'm not 100% happy with this yet because `random_agent` now has 2 allocations. Not sure why. It's still faster (even if just a bit) than the original `random_agent` but the allocation is annoying.

**Comment by Datseris** [comment]
This is because you do not use `$` in the `@benchmark`. Instead of `@benchmark f(a, b)` you should write `@benchmark f($a, $b)`.

**Comment by jacobusmmsmit** [comment]
I think the key reason why it's faster is because `shuffle!` has to generate a permutation for the whole length of the vector every single time, whereas the new default implementation only generates `x` random numbers each time where `x <= length(vector)`.

**Comment by jacobusmmsmit** [comment]
Here are the benchmark results for a `Dict` ABM (the most common case) when `n_agents = 100` and $\mathbb{P}(match) = \frac{1}{3}$:
```julia
# Optimistic
BenchmarkTools.Trial: 10000 samples with 231 evaluations.
 Range (min … max):  398.268 ns … 735.571 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     483.048 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   484.860 ns ±  30.025 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                     ▂▂▂▄▅▆▇▆▅█▇▇▇▇▅▅▃▃▂▂▁                       
  ▁▁▁▁▁▂▂▂▂▂▃▃▄▄▅▆▇▇▇█████████████████████▇▇▆▆▆▄▄▄▃▃▂▂▂▂▂▂▁▁▂▁▁ ▄
  398 ns           Histogram: frequency by time          572 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

# Allocating
BenchmarkTools.Trial: 10000 samples with 201 evaluations.
 Range (min … max):  412.517 ns …  22.619 μs  ┊ GC (min … max): 0.00% … 97.78%
 Time  (median):     476.990 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   503.189 ns ± 883.449 ns  ┊ GC (mean ± σ):  7.62% ±  4.24%

      ▁▆█▆▂                 ▃▃▂▁                                 
  ▁▁▂▄█████▆▃▂▂▂▁▁▁▁▁▁▁▁▂▃▆▇████▇▆▄▄▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  413 ns           Histogram: frequency by time          573 ns <

 Memory estimate: 896 bytes, allocs estimate: 1.

# New random_agent (with optimistic=false)
BenchmarkTools.Trial: 10000 samples with 199 evaluations.
 Range (min … max):  413.945 ns …  23.040 μs  ┊ GC (min … max): 0.00% … 97.67%
 Time  (median):     477.181 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   504.668 ns ± 902.871 ns  ┊ GC (mean ± σ):  7.76% ±  4.24%

      ▂██▅▃                ▂▃▄▂▁                                 
  ▁▁▂▅█████▆▃▂▂▂▁▁▁▁▁▁▁▁▃▄▇█████▆▅▄▄▃▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  414 ns           Histogram: frequency by time          576 ns <

 Memory estimate: 896 bytes, allocs estimate: 1.

# Old random_agent
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.383 μs … 439.625 μs  ┊ GC (min … max): 0.00% … 99.21%
 Time  (median):     1.558 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.610 μs ±   4.382 μs  ┊ GC (mean ± σ):  2.71% ±  0.99%

           ▅▂▃▃▆  ▃ ▁▅▂▂█▂▁ ▁                                  
  ▁▁▁▂▂▃▆▆███████▇███████████▆▄▅▂▂▃▂▁▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  1.38 μs         Histogram: frequency by time        1.95 μs <

 Memory estimate: 896 bytes, allocs estimate: 1.
```

File: src/core/model_abstract.jl [hunk 1]
```diff
@@ -149,24 +149,44 @@ Return a random agent from the model.
...
+    else
+        return allocating_random_agent(model, condition)
     end
-    return a
 end
 
+function allocating_random_agent(model, condition)
+    ids = collect(allids(model))
+    while !isempty(ids)
+        index_id = rand(abmrng(model), eachindex(ids))
```
[the remark this answers is not in the document]
Comment (jacobusmmsmit) [reply src/core/model_abstract.jl:173]: Honestly idk how multiline reviews work (or if they even do) either

File: src/core/model_abstract.jl [hunk 2]
```diff
@@ -149,24 +149,44 @@ Return a random agent from the model.
...
+        condition(model[id]) && return model[id]
+        ids[index_id], ids[end] = ids[end], ids[index_id]
+        pop!(ids)
+    end
+    return nothing
+end
 
+function optimistic_random_agent(model, condition; n_attempts = 3*nagents(model))
+    for _ in 1:n_attempts
+        idx = rand(abmrng(model), allids(model))
```
Comment (Tortar) [inline src/core/model_abstract.jl:185]:
```suggestion
    rng = abmrng(model)
    for _ in 1:n_attempts
        idx = rand(rng, allids(model))
```

**Comment by Tortar** [comment]
Sorry for the inexperience, but we should add it to the changelog under 5.7 and then change 5.7.0 in 5.7.1 in project.toml? Or put it under a new 5.8 and then change 5.7.0 in 5.8.0?

**Comment by jacobusmmsmit** [comment]
Yeah I'm not confident how to handle it properly according to this repo. I'll leave it to George :)

**Review by Tortar — APPROVED** [review]

**Comment by Datseris** [comment]
Thanks guys! Alright, so, performance improvements increment _patch_ version. So this would go under 5.7.1. But now I'm merging as is, because there are other PRs going on, so there is no pressing need to change project.toml version in this PR. Bceause there are a lot of PRs, it is likely that we will release 5.8 anyways, so this should get a changelog entry under 5.8. 

(I'm not waiting for the changelog entry. merging away now, feel free to do followup PR for changelog)

**Comment by Datseris** [comment]
Oh what am I saying. This has to increment the **minor version**, not patch. Because it introduces a new API name, a new keyword for the function. So it anyways goes under 5.8...

**Pull request merged**